frontend/src/components/elements/CodeBlock/CopyButton.test.tsx

```diff
@@ -62,10 +61,11 @@ describe("CopyButton Element", () => {
 
       wrapper.unmount()
 
-      const mockClipboard = Clipboard.mock.instances[0]
-      const mockDestroy = mockClipboard.destroy
+      // const mockClipboard = Clipboard.mock.instances[0]
+      // const mockDestroy = mockClipboard.destroy
 
-      expect(mockDestroy).toHaveBeenCalled()
+      // does not work with useEffect. need to check better way
+      // expect(mockDestroy).toHaveBeenCalled()
```

Comment:
It took me a while, but I understand why this is failing. It comes from [Stale closures](https://dmitripavlutin.com/react-hooks-stale-closures/)

Essentially, on mount, the cleanup function provided has a reference to the null clipboard. Even if the clipboard changes, it doesn't reflect that. Based on the component, if the text changes, it will update the clipboard, which you can verify in the tests by `wrapper.setProps({ text: 'Foo' })` before unmounting.

But that doesn't make sense for this test. Turns out, we depend on how the clipboard works in the `useEffect` but it doesn't reflect it. We actually need to create two effects.

```typescript
  useEffect(() => {
    const node = button.current
    if (node !== null) {
      setClipBoard(new Clipboard(node))
    }
  }, [text])

  useEffect(() => () => {
    if (clipboard !== null) {
      clipboard.destroy()
      setClipBoard(null)
    }
  }, [clipboard])
```

TIL

Reply:
@kmcgrady , since both useEffects would run on mount, won't the clipboard get destroyed as soon as it is created?
Effectively, we do not have a clipboard.
The return function from a useEffect allows one to do cleanup, which is the desired effect right?

Reply:
The second `useEffect` is returning a function above (it's a syntactic handling of the function. Perhaps it should be more clear). That cleanup runs on unmount...How I believe it should work is:

1. Both `useEffects` run on mount.
2. First `useEffect` sets clipboard. That triggers the second `useEffect`.
3. Second `useEffect` pretty much "updates" the clean up function with the latest closure reference to `clipboard`.

When `text` changes:
1. First `useEffect` sets clipboard. That triggers the second `useEffect`.
2. Second `useEffect` pretty much "updates" the clean up function with the latest closure reference to `clipboard`.

Reply:
My bad. Didn't see the lambda in second. Yes, this makes sense. Let me try run the test with this change.

Reply:
The tests are running now. @kmcgrady  thanks for this insight!  Learned a new thing today about stale closures.

Reply:
I think for the sake of simplicity, we should just revert the `CopyButton` to its class component for now.

Reply:
Yes. Will revert CopyButton